### DISCUSSION TITLE
Idiomatic way to remove/search/change if it has a certain value

### DISCUSSION THREAD

**Discussion by crankydillo** [openrewrite/rewrite#4524] [Q&A] [answered]
I need to remove/change a property only if it has a certain value.  I think writing an imperative recipe for this won't be that bad, but just making sure there is no current standard way on this.  Everything I see in properties just searches by key.  [Example](https://docs.openrewrite.org/recipes/properties/search/findproperties)

**Comment by timtebeek** [comment, marked as answer]
You could add a precondition using org.openrewrite.properties.ChangePropertyValue, to match the old value and change it to something else.. Only if that produces any changes would the recipeList containing your removal be executed. Hope that's clear and helps!

**Reply by crankydillo** [reply]
Interesting.  So `ChangePropertyValue` when used as a Precondition won't actually change the property value in the file?

Got to be honest, I never would have thought to do that:)  Is this approach documented or widely used?

**Reply by timtebeek** [reply]
Indeed preconditions never change the actual code; it's what makes this pattern work. :)
Not sure if we've explicitly documented that pattern, but it's certainly helpful here.

**Reply by crankydillo** [reply]
I was looking at some of the code around Preconditions yesterday.  Does what you say work primarily because of the fact that it will not change things and this code:

```java
@Override                                                                                                                                                                                                        
   public @Nullable Tree visit(@Nullable Tree tree, ExecutionContext ctx) {                                                                                                                                         
         Tree t = p.visit(tree, ctx);                                                                                                                                                                                 
         preconditionApplicable = t != tree;                                                                                                                                                                          
         return tree;                                                                                                                                                                                                 
}   
```

`ChangePropertyValue#getVisitor` will generate a different `Tree` in order to actually change the value, so `preconditionApplicable` will be true?

**Reply by timtebeek** [reply]
I'm on mobile, but yes that sounds like it to me. It's also why adding a marker works
